### ISSUE TITLE
ec2.py incorrectly returns empty output

### ISSUE THREAD

**Issue by jsgoller1** [ansible/ansible#19276]
##### ISSUE TYPE
Bug Report

##### COMPONENT NAME
[ec2.py](https://raw.githubusercontent.com/ansible/ansible/devel/contrib/inventory/ec2.py)

##### ANSIBLE VERSION
```
ansible 2.0.0.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```

##### CONFIGURATION
Default ansible.cfg, ec2.ini, and ec2.py in `~`
`sudo apt-get install ansible python-pip`
`pip install boto`
`pip install six`

##### OS / ENVIRONMENT
Ubuntu 16.04

##### SUMMARY
I have been using Terraform to spin up AWS infrastructure; I have multiple instances, ELBs, and other items running in US-East-1. After installation steps listed above, I took my AWS_ACCESS_KEY_ID and AWS_SECRET_KEY_ID and exported them to my environment per the documentation, then ran `~/ec2.py --list`; it waits for between 10-15 seconds, and then returns an empty `hostvars` dictionary.

##### STEPS TO REPRODUCE
Install packages and wget config files as described in CONFIGURATION. Then (replacing YOUR_AWS_API_KEY and YOUR_AWS_API_SECRET_KEY with actual ones from IAM),
```
$ export AWS_ACCESS_KEY_ID='YOUR_AWS_API_KEY'
$ export AWS_SECRET_ACCESS_KEY='YOUR_AWS_API_SECRET_KEY'
$ ~/ec2.py --list
```
##### EXPECTED RESULTS
Standard output when instances and other AWS resources exist. I have six instances that currently exist, called `kubernetes_controller_1`, `kubernetes_worker_1`, `kubernetes_controller_2`, and so on.

##### ACTUAL RESULTS
```
joshua@computer:~$ ./ec2.py 
{
  "_meta": {
    "hostvars": {}
  }
}
```

**Comment by jsgoller1** [comment]
Issue is solved.

**TL;DR**
Had previously installed ansible with Aptitude, not with PIP; also, I didn't have the `cryptography` python module installed.

**Solution**
(assumes you already have ec2.py and ec2.ini in the directory, and your AWS credentials are in ec2.ini)
```
apt-get install apt-get install build-essential libssl-dev libffi-dev python-dev ## needed for cryptography
virtualenv ./venv/
source ./venv/bin/activate
pip install ansible boto six cryptography
./ec2.py --list  ## should now return correct output
```

**Comment by marcinkubica** [comment]
@Jsgoller1 `cryptography` libs are not solution to this. Hitting the problem randomly too. Especially when playing with `regions` in `ec2.ini` file. It can say a region is invalid or AWS down, or returns empty list.

**Comment by volodymyrrudyi** [comment]
+1

**Comment by donkeyx** [comment]
+1

**Comment by ghost** [comment]
Googling has brought me to this issue more than one time when searching for empty host var errors. If your instances are in a private subnet then you will want to double check your ec2.ini file to make sure that the vpc_destination_variable is set to `private_ip_address` rather than `ip_address`

please see below warning provided in the ec2.ini documentation:

WARNING: - instances that are in the private vpc, _without_ public ip address
will not be listed in the inventory until You set:
vpc_destination_variable = private_ip_address

**Comment by jpjaatin** [comment]
Playing with ec2.ini caused me the same issue. Refreshing the cache helped:
`./ec2.py --list --refresh-cache`

**Comment by kbreit** [comment]
I am having this same problem. I set `vpc_destination_variable` to `private_ip_address` and ran with the `--refresh-cache` argument. My `regions` preference is set to `all` at this time.
